### PR TITLE
Allow input parameters in ESOH model

### LINKED ISSUE #2981: [Bug]: `calc_esoh` does not take input parameters
### PyBaMM Version

23.4.1

### Python Version

3.9.16

### Describe the bug

Currently, the `calc_esoh` option of the simulation does not take into account input parameters. Therefore, if there is an input parameter in the OCVs the solver throws an error.

### Steps to Reproduce

```python3
import pybamm

model = pybamm.lithium_ion.DFN()
parameter_values = pybamm.ParameterValues("Chen2020")

def graphite_LGM50_ocp_Chen2020(sto):
    a = pybamm.Parameter("a")
    u_eq = a * (
        1.9793 * pybamm.exp(-39.3631 * sto)
        + 0.2482
        - 0.0909 * pybamm.tanh(29.8538 * (sto - 0.1234))
        - 0.04478 * pybamm.tanh(14.9159 * (sto - 0.2769))
        - 0.0205 * pybamm.tanh(30.4444 * (sto - 0.6103))
    )

    return u_eq

parameter_values.update({
        "Negative electrode OCP [V]": graphite_LGM50_ocp_Chen2020,
})
parameter_values.update({"a": "[input]"}, check_already_exists=False)
experiment = pybamm.Experiment(["Discharge at 1C until 2.5 V"])

sim = pybamm.Simulation(model, parameter_values=parameter_values, experiment=experiment)

sim.solve([0, 3600], inputs={"a": 1})
```
crashes, but changing the last line for
```python3
sim.solve([0, 3600], inputs={"a": 1}, calc_esoh=False)
```
works.

### Relevant log output

```shell
Traceback (most recent call last):
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 260, in solve
    sol = self._solve_full(inputs, ics)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 337, in _solve_full
    sol = sim.solve([0], inputs=inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/simulation.py", line 649, in solve
    self._solution = solver.solve(self.built_model, t_eval, **kwargs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/base_solver.py", line 726, in solve
    model_inputs_list = [
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/base_solver.py", line 727, in <listcomp>
    self._set_up_model_inputs(model, inputs) for inputs in inputs_list
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/base_solver.py", line 1337, in _set_up_model_inputs
    raise pybamm.SolverError(f"No value provided for input '{name}'")
pybamm.expression_tree.exceptions.SolverError: No value provided for input 'a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 264, in solve
    sol = self._solve_split(inputs, ics)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 344, in _solve_split
    x100_sol = x100_sim.solve([0], inputs=inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/simulation.py", line 649, in solve
    self._solution = solver.solve(self.built_model, t_eval, **kwargs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/base_solver.py", line 726, in solve
    model_inputs_list = [
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/base_solver.py", line 727, in <listcomp>
    self._set_up_model_inputs(model, inputs) for inputs in inputs_list
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/base_solver.py", line 1337, in _set_up_model_inputs
    raise pybamm.SolverError(f"No value provided for input '{name}'")
pybamm.expression_tree.exceptions.SolverError: No value provided for input 'a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/expression_tree/input_parameter.py", line 79, in _base_evaluate
    input_eval = inputs[self.name]
KeyError: 'a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/brosaplanella/support-pybamm/mwe_cacl_esoh", line 26, in <module>
    sim.solve([0, 3600], inputs={"a": 1}, calc_esoh=True)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/simulation.py", line 838, in solve
    cycle_sol = pybamm.make_cycle_solution(
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/solution.py", line 809, in make_cycle_solution
    cycle_summary_variables = _get_cycle_summary_variables(cycle_solution, esoh_solver)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/solvers/solution.py", line 873, in _get_cycle_summary_variables
    esoh_sol = esoh_solver.solve(inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 267, in solve
    self._check_esoh_feasible(inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 426, in _check_esoh_feasible
    self.OCV_function.evaluate(inputs={"x": x0_min, "y": y0_max})
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/expression_tree/binary_operators.py", line 107, in evaluate
    right = self.right.evaluate(t, y, y_dot, inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/expression_tree/binary_operators.py", line 106, in evaluate
    left = self.left.evaluate(t, y, y_dot, inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/expression_tree/binary_operators.py", line 106, in evaluate
    left = self.left.evaluate(t, y, y_dot, inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/expression_tree/symbol.py", line 747, in evaluate
    return self._base_evaluate(t, y, y_dot, inputs)
  File "/home/brosaplanella/support-pybamm/env/lib/python3.9/site-packages/pybamm/expression_tree/input_parameter.py", line 82, in _base_evaluate
    raise KeyError("Input parameter '{}' not found".format(self.name))
KeyError: "Input parameter 'a' not found"
```

### PR DESCRIPTION
**Pull request by abillscmu**
# Description

Allows input parameters in the ESOH model. This allows for use of the `initial_soc` keyword in `sim.solve` for variables such as `Positive electrode active material volume fraction`, which previously errored.

Fixes #2981 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/models/full_battery_models/lithium_ion/electrode_soh.py
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:608]: this should create a new dictionary instead of updating inputs in place which might cause bugs elsewhere

File: pybamm/models/full_battery_models/lithium_ion/electrode_soh.py
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:610]: as above

File: pybamm/models/full_battery_models/lithium_ion/electrode_soh.py
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:735]: don't update in place

File: pybamm/models/full_battery_models/lithium_ion/electrode_soh.py
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:742]: don't update in place

File: pybamm/models/full_battery_models/lithium_ion/electrode_soh_half_cell.py
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/electrode_soh_half_cell.py:158]: don't update in place

File: tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_electrode_soh.py
Comment (tinosulzer) [inline tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_electrode_soh.py:247]: don't update in place

File: tests/unit/test_simulation.py
Comment (tinosulzer) [inline tests/unit/test_simulation.py:240]: this test is pretty complicated, do we really need it?

File: pybamm/models/full_battery_models/lithium_ion/electrode_soh_half_cell.py
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/electrode_soh_half_cell.py:31]: IMO better to keep this as the input parameter

**Review by tinosulzer — CHANGES_REQUESTED** [review]

File: pybamm/models/full_battery_models/lithium_ion/electrode_soh_half_cell.py
Comment (abillscmu) [reply pybamm/models/full_battery_models/lithium_ion/electrode_soh_half_cell.py:31]: Really I think these are all use-cases for PsuedoParameters

File: tests/unit/test_simulation.py
Comment (abillscmu) [reply tests/unit/test_simulation.py:240]: A parameter in the OCV function is a slightly different edge case than a parameter elsewhere involved in this model. I think it should be tested. Happy to simplify it a bit though.

**Review by tinosulzer — APPROVED** [review]
Thanks, looks good to me

**Comment by tinosulzer** [comment]
Can you fix coverage?

**Comment by abillscmu** [comment]
> Can you fix coverage?

will do it tomorrow. I'm guessing that line cannot be accessed or should not be accessed.

File: tests/unit/test_simulation.py
Comment (kratman) [inline tests/unit/test_simulation.py:237]: Can we make this its own test rather than putting it inside an existing test?

I also don't see an assert after ```sim.solve(...)``` then it goes onto the next block

File: tests/unit/test_simulation.py
Comment (kratman) [inline tests/unit/test_simulation.py:275]: Similar to above, `test_solve_with_initial_soc` is quite long, can this be moved to its own named test?

**Review by kratman — CHANGES_REQUESTED** [review]

**Comment by kratman** [comment]
> > Can you fix coverage?
> 
> will do it tomorrow. I'm guessing that line cannot be accessed or should not be accessed.

Breaking things into new functions tends to make it easy to cover those safety lines. Generally if you write it in a "functional" style (i.e. constant arguments into the function, returns the values, no state changes to member variables) then it gets trivial to test.

**Comment by abillscmu** [comment]
> > > Can you fix coverage?
> > 
> > 
> > will do it tomorrow. I'm guessing that line cannot be accessed or should not be accessed.
> 
> Breaking things into new functions tends to make it easy to cover those safety lines. Generally if you write it in a "functional" style (i.e. constant arguments into the function, returns the values, no state changes to member variables) then it gets trivial to test.

Taking a look through the code, this case (`known_value not in ["cell capacity", "cyclable lithium inventory"]`) isn't caught anywhere else. It's pretty far outside of the scope of this PR, so I'm going to delete the extra line and move on.

This function is only 13 lines of code, do you really think it should be broken out into multiple separate functions? Or am I misunderstanding something

**Comment by kratman** [comment]
> This function is only 13 lines of code, do you really think it should be broken out into multiple separate functions? Or am I misunderstanding something

I did not look at the whole function, recommending methods for testing error handling. I will say that in general it is best to end an if-elif chain with an else. It lets future developers know that no cases were forgotten and that the intention was to end the chain that way. So my preference would be leaving the else but adding a test for it

**Comment by abillscmu** [comment]
> > This function is only 13 lines of code, do you really think it should be broken out into multiple separate functions? Or am I misunderstanding something
> 
> I did not look at the whole function, just recommending methods for testing error handling. I will say that in general it is best to end an if-elif chain with an else. It lets future developers know that no cases were forgotten and that the intention was to end the chain that way. So my preference would be leaving the else but adding a test for it

I agree, hence why I added the `else` in the first place, I just think it's a separate issue from this PR given that the chain in question occurs quite a few times throughout pybamm without ever having an `else`.

**Comment by kratman** [comment]
@abillscmu Cleanup as we go is usually a good thing, but if you don't want to do it here then can you make a ticket for it with a couple examples of places it appears?

You can assign it to me and I will refactor it later

**Comment by rtimms** [comment]
There are some other changes that would be good to make to the eSOH stuff (see https://github.com/pybamm-team/PyBaMM/issues/3930), so happy to make that ticket into a place to track all the refactoring needed.

**Comment by tinosulzer** [comment]
I am happy to merge this PR once @kratman approves it

**Comment by kratman** [comment]
@tinosulzer All I was waiting for was a ticket for the places where @abillscmu found the missing else statements. Since he already knew where they were

**Comment by abillscmu** [comment]
They really aren't hard to find, all of them I know of are in `pybamm/models/full_battery_models/lithium_ion/electrode_soh.py`

I guess my comment
> the chain in question occurs quite a few times throughout pybamm without ever having an else.

caused some confusion -- "throughout pybamm" lol is a bit of a stretch.

In any case, I added specific exceptions to a few of the higher-level occurrences of this chain that should catch all of them. Any more will likely require a more in-depth refactor, and anyone completing the refactor should have no issues finding the chains (assuming they are still there after the refactor).

**Comment by tinosulzer** [comment]
Looks like tests are related to #3530 being merged in so you'll need to merge develop and fix conflicts

**Pull request merged**
